sos/plugins/lxd.py

```diff
@@ -17,6 +17,7 @@ class LXD(Plugin, UbuntuPlugin):
     plugin_name = 'lxd'
     profiles = ('container',)
     packages = ('lxd',)
+    files = ('/snap/bin/lxd')
```

Comment:
Since `/snap/bin` is being added as a path, I think this should actually be:

```
commands = ('lxd',)
```

Reply:
If the Ubuntu policy is extended so that the `PackageManager` instance reflects the presence of snap packages then there's no need for a command or file trigger - as long as the snap package is called "lxd", or something else we can add to the packages tuple then we can just enable the plugin on that basis. It shouldn't be too hard to merge the package lists from `dpkg` and snap here.

Reply:
I agree it would be the ideal scenario. I think last time I checked a year ago and one of the challenge was that the PackageManager doesn't support detection of >= 2 packages type at the same time. It's either it will work with SNAP only or DEB only, but not both in the same time iirc. There was also other challenges due to the fact that SNAP doesn't behave like DEB nor RPM, meaning SNAP would need is own customise flavor of PackageManager.

Reply:
The `PackageManager` is a really lightweight abstraction - it doesn't impose any such restrictions. The main thing it does is maintain a dictionary of package name/version/release data:

```python
    packages[name] = {'name': name, 'version': version, 'release': release}
```

There are optional hooks for `query_command`, `verify_command` and `files_command`, but implementations aren't obliged to use those - they are just a quick way of bootstrapping a new package manager type without having to write too much code (e.g. I've considered going back to using the Python RPM bindings in the Red Hat policies for performance reasons - the round trip via `rpm -qa` to text and back is quite costly on systems with many packages installed).

A combined dpkg/snap `PackageManager` would just need to override `get_pkg_list()` to obtain NVR data from both and add those to the `packages` dictionary.

Reply:
(also note that `PackageManager` support isn't a prerequisite for the suggestion @TurboTurtle made - evaluation of `commands` triggers just relies on the ability to search `PATH` - see `sos.utilities.is_executable()`).

Reply:
@bmr-cymru just to make sure we are on the same page. Do you want me to explore the get_pkg_list() to obtain NVR data from both approach or use the `commands` approach @TurboTurtle suggested ?

Reply:
I'm just saying that both are valid ways of triggering the plugin and are preferable to a hard coded absolute path in the plugin `files` list - implementing `PackageManager` support for snap (and doing something similar for the other auxiliary package managers in the policy classes for other distros) has some obvious advantages but is a bigger piece of work to undertake.

Reply:
I totally agree that in the long run we would want a `PackageManager` support for snap (and even for flatpak)

What do you think if for now if I change the PR for `commands`, but will consider the former in my next set of changes to submit in sosreport. That's definitely something I want to pursue for quite some time. My previous attempts weren't successful and I unfortunately didn't had much time to resume the work on it but will do in 2020.

Does that look like a good plan for you ?

Reply:
For now we want to cover snap specifically for Canonical products only : lxd, juju, canonical-livepatch and CDK (Canonical Distribution of K8s). Other than that no other plugins should pay attention to snap for now, but in the long run yes they might so the `PackageManager` approach will be ideal before we cross that bridge, but we still have time. If for now we can cover Canonical product while I'm thinking/working about the ultimate snap pkg support, it would be good.

Reply:
@BryanQuigley thoughts ?